### PR TITLE
refactor: add "valid-as-of-block" in reply when getting txs to forge

### PR DESCRIPTION
**Pull request by vasild**
When getting the top transactions from the pool with the intention to
forge them in the next block we don't want already forged transactions
to be in that list. However, the pool cannot guarantee that there will
not be forged transactions because of the nature of forging - new blocks
are forged every 8 seconds. So even if the reply from the pool is
"correct" (it does not contain forged transactions), then it could happen
that 0.1 seconds after that some of those transactions get forged.

We can guarantee that the list of returned transactions does not contain
forged ones as of some block height (H). That is - none of the returned
transactions is forged in block with height H or earlier blocks. So we
append H to the list of transactions in the reply. The forger knows the
height of the block he is forging and can decide if the reply suits him
or not. This patch changes the forger to retry a few times if the reply
is outdated and if still outdated - try to forge a block nevertheless.

This fixes https://github.com/ArkEcosystem/core/issues/1402

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (improve a current implementation without adding a new feature or fixing a bug)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Build (changes that affect the build system)
- [ ] Docs (documentation only changes)
- [ ] Test (adding missing tests or fixing existing tests)
- [ ] Other... Please describe:

## Checklist

- [ ] I have read the [CONTRIBUTING](https://docs.ark.io/developers/guidelines/contributing.html) documentation
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by kristjank** [comment]
Just had another idea. So forgers know when is his turn. Befor his turn, when the previous forger is going to forge, forger already sends a wake call to the node. We have enough time 12s in this window to also call removeForgedTransactions. In this way we could then completely remove the call from getTransactions, as they would be removed before the actual forging starts.

Then the block of previous delegate is received and its transactions removed from pool.

File: packages/core-blockchain/lib/blockchain.js [hunk 1]
```diff
@@ -528,20 +528,28 @@ module.exports = class Blockchain {
...
    */
-  getUnconfirmedTransactions(blockSize) {
-    const transactions = this.transactionPool.getTransactionsForForging(
-      blockSize,
-    )
+  async getUnconfirmedTransactions(n, asOfHeight = undefined) {
+    let transactions = this.transactionPool.getTransactions(0, n)
+    const transactionsIds = transactions.map(t => t.id)
+
+    const forgedIds = await this.database.getForgedTransactionsIds(transactionsIds, asOfHeight)
```
Comment (kristjank) [inline packages/core-blockchain/lib/blockchain.js]: change to lastForged Height where this forger has succesfully forged. 
`asOfHeight=lastForgedBlockByDelegate.height`

File: packages/core-forger/lib/manager.js [hunk 2]
```diff
@@ -202,7 +202,7 @@ module.exports = class ForgerManager {
     //
     // this.client.emitEvent('forger.started', delegate.publicKey)
 
-    const transactions = await this.__getTransactionsForForging()
+    const transactions = await this.__getTransactionsForForging(round.lastBlock.height)
```
Comment (kristjank) [inline packages/core-forger/lib/manager.js]: same comment here - maybe round.lastForgedBlock by delegate

**Comment by faustbrian** [comment]
@vasild This can now be adjusted to TypeScript on the `develop` branch.

**Comment by faustbrian** [comment]
@kristjank do we close this as was discussed to look for a simpler solution?

**Comment by faustbrian** [comment]
Will close this for, can revisit it later on if needed as simpler solutions were discussed.

**Pull request closed without merging**